src/shared_types.rs

```diff
@@ -115,6 +115,19 @@ impl Identifier {
     }
 }
 
+impl FromStr for Identifier {
+    type Err = ErrorKind;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let id = Self(s.into());
+        if id.is_valid() {
+            Ok(id)
+        } else {
+            Err(ErrorKind::BadIdentifier)
+        }
+    }
+}
```

Comment:
I think this is the right approach, but we should be more rigorous.

- The inner string on `Identifier` should be private; you should only be able to construct an identifier via some function that performs validation
- `is_valid` should move out of the `Identifier` impl and just be part of `from_str`?
- `Identifier` could get a constructor, `fn new(&str) -> Result<Self, ErrorKind>` that just calls this `from_str` impl under the hood?
- Everywhere that we currently construct an identifier directly should be changed to use this constructor

This can be a follow-up PR (or I can do it) if that makes the most sense?

Reply:
Sounds good, yes :thinking: I think I used is_valid for something but it can be a freestanding function in the Impl. I'll mull over Pens some more, so fel free to toss it in via PR :)